compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -1329,17 +1329,15 @@ private boolean isValidAnnotationType(BType type) {
         }
 
         switch (type.tag) {
-            case TypeTags.RECORD:
-//                BRecordType recordType = (BRecordType) type;
-//                return recordType.fields.stream().allMatch(field -> types.isAnydata(field.type)) &&
-//                        (recordType.sealed || types.isAnydata(recordType.restFieldType));
             case TypeTags.MAP:
-//                return types.isAnydata(((BMapType) type).constraint);
-                return true;
+            case TypeTags.RECORD:
+                return types.isAssignable(type, symTable.anydataOrReadOnlyMapType);
             case TypeTags.ARRAY:
                 BType elementType = ((BArrayType) type).eType;
-                return (elementType.tag == TypeTags.MAP || elementType.tag == TypeTags.RECORD) &&
-                        isValidAnnotationType(elementType);
+                if (elementType.tag == TypeTags.MAP || elementType.tag == TypeTags.RECORD) {
+                    return types.isAssignable(elementType, symTable.anydataOrReadOnlyMapType);
+                }
+                return false;
```

Comment:
Instead, should we do 
```suggestion
                return isValidAnnotationType(elementType);
```